Modules/_struct.c

```diff
@@ -1930,15 +1930,40 @@ s_pack_into(PyObject *self, PyObject **args, Py_ssize_t nargs, PyObject *kwnames
         return NULL;
     }
 
+    /* Check that negative offset is low enough to fit data */
+    if (offset < 0 && offset + soself->s_size > 0) {
+        PyErr_Format(StructError,
+                     "pack_into requires negative offset not higher than %zd "
+                     "for packing %zd bytes",
+                     -soself->s_size,
+                     soself->s_size);
+        PyBuffer_Release(&buffer);
+        return NULL;
+    }
+
+    /* Check that negative offset is not crossing buffer boundary */
+    if (offset < 0 && offset + buffer.len < 0) {
+        PyErr_Format(StructError,
+                     "pack_into requires negative offset not lower than %zd "
+                     "for packing into buffer of %zd bytes",
+                     -buffer.len,
+                     buffer.len);
+        PyBuffer_Release(&buffer);
+        return NULL;
+    }
+
     /* Support negative offsets. */
     if (offset < 0)
         offset += buffer.len;
 
     /* Check boundaries */
-    if (offset < 0 || (buffer.len - offset) < soself->s_size) {
+    if ((buffer.len - offset) < soself->s_size) {
         PyErr_Format(StructError,
-                     "pack_into requires a buffer of at least %zd bytes",
-                     soself->s_size);
+                     "pack_into requires a buffer of at least %zd bytes for "
+                     "packing %zd bytes at offset %zd",
+                     soself->s_size + (offset > 0 ? offset : 0),
```

Comment:
I think it would be more helpful to output an actual size of the buffer rather of `soself->s_size + offset`. The user can add two reported numbers.